source/Calamari.Aws/UploadAwsS3Command.cs

```diff
@@ -29,68 +28,76 @@ namespace Calamari.Aws
     public class UploadAwsS3Command : AwsCommand
     {
-        readonly IAmazonS3 amazonS3Client;
         readonly IProvideS3TargetOptions optionsProvider;
         readonly ICalamariFileSystem fileSystem;
         readonly ISubstituteInFiles substituteInFiles;
         readonly IExtractPackage extractPackage;
-        readonly string bucketName;
-        readonly S3TargetMode s3TargetMode;
-        readonly bool isMd5HashSupported;
+        readonly Lazy<Task<IAmazonS3>> amazonS3Cleint;
 
         static readonly HashSet<S3CannedACL> S3CannedAcls = new HashSet<S3CannedACL>(ConstantHelpers.GetConstantValues<S3CannedACL>());
 
+        string bucketName;
+        S3TargetMode s3TargetMode;
+        bool isMd5HashSupported;
+        PathToPackage pathToPackage;
+
         public UploadAwsS3Command(
             ILog log,
             IVariables variables,
-            IAmazonS3 amazonS3Client,
-            IAmazonSecurityTokenService amazonSecurityTokenService,
-            IAmazonIdentityManagementService amazonIdentityManagementService,
+            IAmazonClientFactory amazonClientFactory,
             IProvideS3TargetOptions optionsProvider,
             ICalamariFileSystem fileSystem,
             ISubstituteInFiles substituteInFiles,
-            IExtractPackage extractPackage) : base(log, variables, amazonSecurityTokenService, amazonIdentityManagementService)
+            IExtractPackage extractPackage) : base(log, variables, amazonClientFactory)
         {
-            this.amazonS3Client = amazonS3Client;
             this.optionsProvider = optionsProvider;
             this.fileSystem = fileSystem;
             this.substituteInFiles = substituteInFiles;
             this.extractPackage = extractPackage;
 
-            // TODO: these variable names need to be shared with Sashimi.Aws, but how?
-            bucketName = variables.Get("Octopus.Action.Aws.S3.BucketName")?.Trim();
-            s3TargetMode = GetS3TargetMode(variables.Get("Octopus.Action.Aws.S3.TargetMode"));
-            isMd5HashSupported = HashCalculator.IsAvailableHashingAlgorithm(MD5.Create);
+            amazonS3Cleint = new Lazy<Task<IAmazonS3>>(amazonClientFactory.GetS3Client);
         }
 
-        //TODO: Further refactor is necessary if we have the capacity
-        protected override void Execute(RunningDeployment deployment)
+        public override void Dispose()
         {
-            if (!fileSystem.FileExists(deployment.PackageFilePath))
-                throw new CommandException("Could not find package file: " + deployment.PackageFilePath);
+            if (amazonS3Cleint.IsValueCreated) amazonS3Cleint.Value.Dispose();
 
+            base.Dispose();
+        }
+
+        //TODO: Further refactor is necessary if we have the capacity
+        protected override async Task ExecuteCore()
+        {
+            // TODO: these variable names need to be shared with Sashimi.Aws, but how?
```

Comment:
https://octopusdeploy.slack.com/archives/C24PH407Q/p1589930762203700

Can be done in a separate PR